datasets/ONS_CPA/main.py

```diff
@@ -6,196 +6,63 @@
-    elif gre.fullmatch(GROUP_RE, p):
-        if p[1] == '.':
-            return f"group/0{p}"
-        else:
-            return f"group/{p}"
-    elif gre.fullmatch(CLASS_RE, p):
-        return f"class/{p}"
-    elif gre.fullmatch(ONS_RE, p):
-        return f"ons/{p}"
-    else:
-        assert False, f"no match for {p}"
-    
-cdids['PRODUCT'].cat.categories = cdids['PRODUCT'].cat.categories.map(product2cpa)
-cdids['DIRECTION'].cat.categories = cdids['DIRECTION'].cat.categories.map(
-    lambda x: 'balance' if x == 'BAL' else 'imports' if x == 'IM' else 'exports' if x == 'EX' else None
-)
-cdids.rename(columns={
-    'PRODUCT': 'CPA 2008',
-    'AREA': 'ONS Partner Geography',
-    'DIRECTION': 'Flow',
-    'PRICE': 'Price Classification',
-    'SEASADJ': 'Seasonal Adjustment',
-    'BASIS': 'International Trade Basis'
-}, inplace=True)
-# -
+# Unstack turns the dataframe to a series, creating a complex multiindex for rows
+series = df.unstack()
 
-# Merge in the dimension values
+# Name the series in the dataset, which helps when converting back to a dataframe
+series.name = 'Marker'
 
-observations = observations.merge(
-    cdids, how = 'left', left_on = ['CDID', 'Periodicity'], right_index=True)
-observations['Measure Type'] = pd.Series('GBP Total', index=observations.index, dtype='category')
-observations['Unit'] = pd.Series('gbp-million', index=observations.index, dtype='category')
-observations = observations[['ONS Partner Geography', 'Period', 'CDID', 'International Trade Basis',
-                             'Flow', 'CPA 2008', 'Price Classification', 'Seasonal Adjustment',
-                             'Measure Type', 'Value', 'Unit']].drop_duplicates()
 
 # +
-observations.rename(columns={'Flow':'Flow Directions'}, inplace=True)
+# Convert the series to a dataframe, and move the row multiindex to column values
+df = series.to_frame().reset_index()
 
-#Flow has been changed to Flow Direction to differentiate from Migration Flow dimension
+# Rename the column named 0 to 'Period'
+df.rename({0: 'Period'}, axis=1, inplace=True)
 # -
 
-observations
+# Set period values to proper format
+df['Period'].loc[df['Period'].str.len() == 7] = df['Period'].apply(lambda x:'/id/quarter/{year}-{quarter}'.format(year=x[:4], quarter=x[-2:]))
+df['Period'].loc[df['Period'].str.len() == 4] = df['Period'].apply(lambda x:'/id/year/{year}'.format(year=x[:4]))
+
+# Interprate the release dates and format
+df['Next release'] = pd.to_datetime(df['Next release'], format='%d %B %Y').dt.strftime('/id/day/%Y-%m-%d')
+df['Release Date'] = pd.to_datetime(df['Release Date'], format='%d-%m-%Y').dt.strftime('/id/day/%Y-%m-%d')
+
+# We don't use the data within these cells
+df.drop(labels=['Title'], axis=1, inplace=True)
 
-from pathlib import Path
-out = Path('out')
-out.mkdir(exist_ok=True)
-observations.drop_duplicates().to_csv(out / 'observations.csv', index = False)
+# Add dataframe is in the cube
+cubes.add_cube(scraper, df, info['title'])
```

Comment:
Might be better to extract the title from distribution itself in case of any human error pulled in from info file. e.g replace info["title"] with distribution.title.

Reply:
Done.